packages/SwingSet/src/devices/plugin-src.js

```diff
@@ -0,0 +1,134 @@
+/* global harden HandledPromise */
+
+import { makeCapTP } from '@agoric/captp';
+
+export function buildRootDeviceNode(tools) {
+  const { SO, getDeviceState, setDeviceState, endowments } = tools;
+  const restart = getDeviceState();
+
+  let registeredReceiver = restart && restart.registeredReceiver;
+
+  const senders = {};
+  // Take a shallow copy so that these are not frozen.
+  const connectedMods = restart ? [...restart.connectedMods] : [];
+  const nextEpochs = restart ? [...restart.nextEpochs] : [];
+  const connectedState = restart ? [...restart.connectedState] : [];
+
+  function saveState() {
+    setDeviceState(
+      harden({
+        registeredReceiver,
+        // Take a shallow copy so that these are not frozen.
+        nextEpochs: [...nextEpochs],
+        connectedMods: [...connectedMods],
+        connectedState: [...connectedState],
+      }),
+    );
+  }
+  // Register our first state.
+  saveState();
+
+  /**
+   * Load a module and connect to it.
+   * @param {string} mod module with an exported `bootPlugin(state = undefined)`
+   * @param {number} [index=connectedMods.length] the module instance index
+   * @param {(obj: Record<string, any>) => void} receive a message from the module
+   * @returns {(obj: Record<string, any>) => void} send a message to the module
+   */
+  function connect(mod, index = connectedMods.length) {
+    try {
+      // Allocate this module first.
+      if (connectedMods[index] === undefined) {
+        connectedMods[index] = mod;
+      }
+      if (connectedMods[index] !== mod) {
+        throw TypeError(
+          `Index ${index} is already allocated to ${connectedMods[index]}, not ${mod}`,
+        );
+      }
+      const epoch = nextEpochs[index] || 0;
+      nextEpochs[index] = epoch + 1;
+      saveState();
+
+      const modNS = endowments.require(mod);
+      const receiver = obj => {
+        // console.info('receiver', index, obj);
+
+        // We need to run the kernel after the send-only.
+        endowments.queueThunkForKernel(() =>
+          SO(registeredReceiver).dispatch(index, obj),
+        );
+      };
+      // Create a bootstrap reference from the module.
+      const bootstrap = modNS.bootPlugin(
+        harden({
```

Comment:
to test the transcript replay is working correctly, we should pass through `testLog` here, so the loaded module can write things where the unit test can see them